quinn/src/streams.rs

```diff
@@ -371,10 +371,23 @@ where
         cx: &mut Context,
         buf: &mut ReadBuf<'_>,
     ) -> Poll<Result<(), ReadError>> {
-        self.poll_read_generic(cx, |conn, stream| {
-            conn.inner
-                .read(stream, buf.remaining(), true)
-                .map(|val| val.map(|chunk| buf.put_slice(&chunk.bytes)))
+        self.poll_read_generic(cx, true, |chunks| {
+            let mut read = false;
+            loop {
+                if buf.remaining() == 0 {
+                    return Ok(Some(()));
+                }
+
+                match chunks.next(buf.remaining()) {
+                    Ok(Some(chunk)) => {
+                        buf.put_slice(&chunk.bytes);
+                        read = true;
+                    }
+                    Ok(None) | Err(_) if read => return Ok(Some(())),
+                    Ok(None) => return Ok(None),
+                    Err(e) => return Err(e),
+                }
+            }
```

Comment:
This pattern (and below) can lose errors. Consider what happens when both data and an error are waiting. If the application reads all the data, the error will not be reported, but the stream's state (error included) may then be disposed. The next read call may then see `UnknownStream` rather than the correct error.

Rather than fusing errors in `Chunks`, which is short lived, we should do so here at the high level, where they're guaranteed to be preserved as long as we like.

Reply:
This is done.